cpp/src/datetime/datetime_ops.cu

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/types.hpp>
+#include <cudf/wrappers/timestamps.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/transform.h>
+#include <thrust/functional.h>
+#include <thrust/execution_policy.h>
+
+namespace cudf {
+namespace datetime {
+namespace detail {
+
+  typedef enum {
+    invalid = 0,
+    year,
+    month,
+    day,
+    weekday,
+    hour,
+    minute,
+    second,
+  } datetime_component;
+
+  template <datetime_component DatetimeComponent>
+  struct extract_component_operator {
+
+    column_device_view column;
+
+    CUDA_HOST_DEVICE_CALLABLE extract_component_operator(column_device_view col) : column(col) {}
+
+    template <typename Element>
+    typename std::enable_if_t<not cudf::is_timestamp_t<Element>::value, int16_t>
+    CUDA_HOST_DEVICE_CALLABLE operator()(size_type const i) const { return 0; }
+
+    template <typename Element>
+    typename std::enable_if_t<cudf::is_timestamp_t<Element>::value, int16_t>
+    CUDA_HOST_DEVICE_CALLABLE operator()(size_type const i) const {
+
+      using namespace simt::std::chrono;
+      using Duration = typename Element::duration;
+
+      auto ts = column.element<Element>(i);
+      auto days_since_epoch = floor<days>(ts);
+
+      switch (DatetimeComponent) {
```

Comment:
Have you looked at the generated assembly to see if this works such that the compiler compiles away the rest of the function/cases when instantiated with a particular `datetime_component`?

You can use https://godbolt.org/ for this.

Reply:
Looks like it is in this stripped-down example: https://godbolt.org/z/O9RSY0